Improve file globbing logic (fixes #5057)

mitchellwrosen opened the pull request:
I tested this change manually by running `cabal sdist` on https://github.com/RyanGlScott/cabal-sdist-bug/tree/910078306b4e127996bd7c9208a7e741b9f74351:

```
$ cabal sdist
Building source dist for cabal-sdist-bug-0.1.0...
Source tarball created: dist/cabal-sdist-bug-0.1.0.tar.gz
```

hvr commented:
This likely also needs a `cabal check`, as old cabals will choke.

mitchellwrosen commented:
Sorry, I'm unclear on which came first, these TODOs or your comment about how the current behavior was actually intended. Should I close this PR?

23Skidoo commented:
That behaviour was intended originally, but there seems to be a consensus that we should relax it a bit.

mitchellwrosen commented:
I've added a test and updated `changelog`. Could I have some guidance on these remaining TODOs?

> (& make sure that the documented limitations for * in filenames remain in place).

What are the documented limitations for `*` in filenames?

> possibly a cabal check compat-warning

What does this mean?

> possibly user guide update (to document that starting with version x.y, fileglobbing supports multi-part extensions)

Where is the user guide located?

Thanks!

phadej commented:
I'm really sorry that we (me and @hvr) are demanding. I try to explain:

---

> possibly a cabal check compat-warning`

run `cabal check`. The implementation probably have to be in `Distribution.PackageDescription.Check`.
There is e.g. `check use of "data-files: data/*.txt" syntax`.

The issue is that:
- the change is good for ergonomics
- but if someone else developing the package would use older `cabal-install`, their `cabal sdist` won't pick all the files, resulting in broken package.
    - unfortunately there isn't `cabal check` which checks that globs actually pick at least *some* files. There should be.
    - new version should warn, that 'extended globbing' is needed, and `cabal-version` is too low.
        - that's tricky, but doable

The scenario is not that different to #5073:
if some CI setup (for any reason) uses older `cabal-install` for some job, than `cabal sdist` will produce broken tarballs, resulting in confusingly failing builds.

> Where is the user guide located?

`Cabal/doc`
